### ISSUE TITLE
out effector wastes the heap

### ISSUE THREAD

**Issue by jrte** [jrte/ribose#19]
It is decoding parameter values and writing Unicode to stdout and a number of transient ByteBuffer object are created in the process. That's just wrong. Values should be written directly as utf8 bytes without copying or decoding. 

Test script runs LinuxKernel transducer on a 20MB byte array preloaded in the heap starting with a 22MB heap and java pinned to CPUs 3&4. There the first run is with the out effector disabled. It is enabled for the second run and the heap just keeps on growing:

```
$ cat local/LinuxKernel-gc 
#! /bin/bash
gclog="-Xlog:gc=trace:file=/tmp/LinuxKernel-gc.log"
start="-Dregex.out.enabled=false -cp jars/ribose-0.0.0.jar -Dfile.encoding=UTF-8 -classpath jars/ribose-0.0.0.jar com.characterforming.ribose"
input="test-patterns/inputs/kern-10.log"
model="build/patterns/BaseTarget.model"
java="/usr/bin/time numactl --physcpubind 3,4 java -Xms22m"

echo "lines words characters = $(wc $input)"
echo "LinuxKernel with output disabled"
$java $gclog -Djrte.out.enabled=false $start.RiboseRuntime \
  --nil LinuxKernel  "$input" "$model" 2>/tmp/LinuxKernel-gc.time
cat /tmp/LinuxKernel-gc.time
cat /tmp/LinuxKernel-gc.log

rm /tmp/LinuxKernel-gc.log
echo "LinuxKernel with output enabled"
$java $gclog -Djrte.out.enabled=true $start.RiboseRuntime \
  --nil LinuxKernel "$input" "$model" 1>/tmp/LinuxKernel.out 2>/tmp/LinuxKernel-gc.time
echo "lines words characters = $(wc /tmp/LinuxKernel.out)"
cat /tmp/LinuxKernel-gc.time
cat /tmp/LinuxKernel-gc.log
```

Output for each run shows time metrics and gc log trace:
```
lines words characters =   115080  2199300 20420730 test-patterns/inputs/kern-10.log
LinuxKernel with output disabled
0.34user 0.06system 0:00.30elapsed 133%CPU (0avgtext+0avgdata 80276maxresident)k
0inputs+72outputs (8major+15283minor)pagefaults 0swaps
[0.002s][trace][gc] MarkStackSize: 4096k  MarkStackSizeMax: 16384k
[0.024s][debug][gc] ConcGCThreads: 1 offset 4
[0.024s][debug][gc] ParallelGCThreads: 2
[0.025s][debug][gc] Initialize mark stack with 4096 chunks, maximum 16384
[0.026s][info ][gc] Using G1
[0.106s][info ][gc] GC(0) Pause Young (Concurrent Start) (G1 Humongous Allocation) 1M->1M(24M) 3.168ms
[0.106s][info ][gc] GC(1) Concurrent Cycle
[0.111s][info ][gc] GC(1) Pause Remark 21M->21M(44M) 0.273ms
[0.114s][info ][gc] GC(1) Pause Cleanup 21M->21M(44M) 0.040ms
[0.114s][info ][gc] GC(1) Concurrent Cycle 8.494ms
LinuxKernel with output enabled
lines words characters =   70279  210837 7683656 /tmp/LinuxKernel.out
1.15user 1.96system 0:02.84elapsed 109%CPU (0avgtext+0avgdata 151680maxresident)k
0inputs+15080outputs (8major+37002minor)pagefaults 0swaps
[0.002s][trace][gc] MarkStackSize: 4096k  MarkStackSizeMax: 16384k
[0.006s][debug][gc] ConcGCThreads: 1 offset 4
[0.006s][debug][gc] ParallelGCThreads: 2
[0.006s][debug][gc] Initialize mark stack with 4096 chunks, maximum 16384
[0.006s][info ][gc] Using G1
[0.076s][info ][gc] GC(0) Pause Young (Concurrent Start) (G1 Humongous Allocation) 1M->1M(24M) 3.144ms
[0.076s][info ][gc] GC(1) Concurrent Cycle
[0.085s][info ][gc] GC(1) Pause Remark 21M->21M(44M) 1.679ms
[0.085s][info ][gc] GC(1) Pause Cleanup 21M->21M(44M) 0.047ms
[0.085s][info ][gc] GC(1) Concurrent Cycle 8.977ms
[0.280s][info ][gc] GC(2) Pause Young (Normal) (G1 Evacuation Pause) 33M->21M(44M) 4.602ms
[0.366s][info ][gc] GC(3) Pause Young (Concurrent Start) (G1 Evacuation Pause) 33M->21M(44M) 3.377ms
[0.366s][info ][gc] GC(4) Concurrent Cycle
[0.368s][info ][gc] GC(4) Pause Remark 21M->21M(44M) 1.098ms
[0.369s][info ][gc] GC(4) Pause Cleanup 21M->21M(44M) 0.059ms
[0.369s][info ][gc] GC(4) Concurrent Cycle 3.194ms
[0.448s][info ][gc] GC(5) Pause Young (Normal) (G1 Evacuation Pause) 33M->21M(44M) 3.622ms
[0.526s][info ][gc] GC(6) Pause Young (Concurrent Start) (G1 Evacuation Pause) 33M->21M(69M) 3.219ms
[0.526s][info ][gc] GC(7) Concurrent Cycle
[0.528s][info ][gc] GC(7) Pause Remark 21M->21M(69M) 0.416ms
[0.529s][info ][gc] GC(7) Pause Cleanup 21M->21M(69M) 0.063ms
[0.529s][info ][gc] GC(7) Concurrent Cycle 2.456ms
[0.669s][info ][gc] GC(8) Pause Young (Normal) (G1 Evacuation Pause) 42M->21M(69M) 3.897ms
[0.836s][info ][gc] GC(9) Pause Young (Normal) (G1 Evacuation Pause) 46M->21M(69M) 3.236ms
[1.011s][info ][gc] GC(10) Pause Young (Normal) (G1 Evacuation Pause) 48M->21M(69M) 3.038ms
[1.193s][info ][gc] GC(11) Pause Young (Normal) (G1 Evacuation Pause) 49M->21M(90M) 3.268ms
[1.427s][info ][gc] GC(12) Pause Young (Normal) (G1 Evacuation Pause) 56M->21M(90M) 5.914ms
[1.686s][info ][gc] GC(13) Pause Young (Normal) (G1 Evacuation Pause) 61M->21M(90M) 3.315ms
[1.960s][info ][gc] GC(14) Pause Young (Normal) (G1 Evacuation Pause) 64M->21M(90M) 3.178ms
[2.254s][info ][gc] GC(15) Pause Young (Normal) (G1 Evacuation Pause) 66M->21M(110M) 10.231ms
[2.599s][info ][gc] GC(16) Pause Young (Normal) (G1 Evacuation Pause) 73M->21M(110M) 5.875ms
```

**Comment by jrte** [comment]
After fix ([cfd8a5f](https://github.com/jrte/ribose/commit/cfd8a5ffa7a7b5a585a80848c54390a8a5b137b7)):

```
$ local/LinuxKernel-gc 
lines words characters =   115080  2199300 20420730 test-patterns/inputs/kern-10.log
LinuxKernel with output disabled
0.34user 0.05system 0:00.30elapsed 130%CPU (0avgtext+0avgdata 80548maxresident)k
0inputs+72outputs (8major+15279minor)pagefaults 0swaps
[0.002s][trace][gc] MarkStackSize: 4096k  MarkStackSizeMax: 16384k
[0.028s][debug][gc] ConcGCThreads: 1 offset 4
[0.028s][debug][gc] ParallelGCThreads: 2
[0.029s][debug][gc] Initialize mark stack with 4096 chunks, maximum 16384
[0.029s][info ][gc] Using G1
[0.107s][info ][gc] GC(0) Pause Young (Concurrent Start) (G1 Humongous Allocation) 1M->1M(24M) 3.290ms
[0.107s][info ][gc] GC(1) Concurrent Cycle
[0.113s][info ][gc] GC(1) Pause Remark 21M->21M(44M) 0.376ms
[0.115s][info ][gc] GC(1) Pause Cleanup 21M->21M(44M) 0.055ms
[0.115s][info ][gc] GC(1) Concurrent Cycle 7.856ms
LinuxKernel with output enabled
lines words characters =   70279  210837 7683656 /tmp/LinuxKernel.out
0.69user 1.88system 0:02.45elapsed 105%CPU (0avgtext+0avgdata 79516maxresident)k
0inputs+15080outputs (8major+15304minor)pagefaults 0swaps
[0.003s][trace][gc] MarkStackSize: 4096k  MarkStackSizeMax: 16384k
[0.027s][debug][gc] ConcGCThreads: 1 offset 4
[0.027s][debug][gc] ParallelGCThreads: 2
[0.027s][debug][gc] Initialize mark stack with 4096 chunks, maximum 16384
[0.028s][info ][gc] Using G1
[0.101s][info ][gc] GC(0) Pause Young (Concurrent Start) (G1 Humongous Allocation) 1M->1M(24M) 3.108ms
[0.101s][info ][gc] GC(1) Concurrent Cycle
[0.114s][info ][gc] GC(1) Pause Remark 21M->21M(44M) 5.737ms
[0.115s][info ][gc] GC(1) Pause Cleanup 21M->21M(44M) 0.065ms
[0.115s][info ][gc] GC(1) Concurrent Cycle 14.303ms
```

**Comment by jrte** [comment]
From gceasy.io:

[Before](https://gceasy.io/my-gc-report.jsp?p=c2hhcmVkLzIwMjIvMDQvMjAvLS1nYy4xLS0yMy0yOC0yNA==&channel=WEB)
![image](https://user-images.githubusercontent.com/24707461/164341638-492d4379-8ac0-40ec-864c-37f11cf98617.png)

[After](https://gceasy.io/my-gc-report.jsp?p=c2hhcmVkLzIwMjIvMDQvMjAvLS1nYy4yLS0yMy0zMC0yMA==&channel=WEB)
![image](https://user-images.githubusercontent.com/24707461/164341849-7298af53-6cc9-4f4b-b6e1-514116ea7d7e.png)
